### PR TITLE
Add basic RoPE implementation to m3 core

### PR DESCRIPTION
**Pull request by bvandermoon**
# Description

Adds `maxtext.m3.core.rope.apply_rope`, the base rotary positional embedding for
m3, as a pure function rather than a module — RoPE holds no state and allocates
no parameters, so there is nothing to place on a mesh. Only the base variant
lives here; specialized variants (YaRN, LLaMA 3.1 extended context,
partial-rotary, mRoPE) belong in the model directories that need them.

Numerics match the legacy `RotaryEmbedding` bit-exactly — half-split ("rotate
half") convention, angles in float32, sinusoids cast to the input dtype before
the rotation — so migrated models reproduce legacy numbers rather than
approximating them.

Two intentional deviations from the design doc sketch:
- `positions` expands to `(batch, length, 1, 1)` so it broadcasts against
  `(batch, length, heads, head_dim)`; the sketch's `positions[..., None]` does
  not broadcast. Mismatched shapes now raise instead of broadcasting silently.
- `min_timescale` is exposed, matching the legacy `rope_min_timescale` config.

Second m3 PR, following the package scaffolding in #5291. Still not wired into
any MaxText entry point.

# Tests

New `tests/unit/m3/rope_test.py` covers RoPE's defining properties
(position-zero identity, norm preservation, relative-position invariance of
attention scores, packed sequences), bit-exact parity against the legacy
`RotaryEmbedding` at float32 and bfloat16, and argument validation.

Run on a v6e VM (`JAX_PLATFORMS=cpu` since these tests are CPU-only and the
conftest skips `cpu_only` tests on accelerator testbeds):

    JAX_PLATFORMS=cpu python3 -m pytest tests/unit/m3/ -q

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable. (N/A — m3 is not wired into an entry point yet.)
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files). (Package docs are inline; `m3/core/__init__.py` updated.)

### CODE REVIEW

File: src/maxtext/m3/core/rope.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  if positions.shape != x.shape[:2]:
+    raise ValueError(
+        f"apply_rope expects positions of shape (batch, length) = {x.shape[:2]}; got shape {positions.shape}."
+    )
+
+  half_dim = head_dim // 2
+
+  # Geometric series of rotation periods, one per dimension pair. Matches the
+  # legacy `RotaryEmbedding.timescale` expression term for term.
+  fraction = 2 * jnp.arange(0, half_dim) / head_dim
```
[the remark this answers is not in the document]
Comment (bvandermoon) [reply src/maxtext/m3/core/rope.py]: Took the redundant 0 removal, thanks. Leaving the arange dtype implicit though: this expression intentionally mirrors legacy `RotaryEmbedding.timescale`, which uses the integer arange, and that's what keeps the two bit-exact. With jax_enable_x64 off — MaxText never enables it — the two forms are bitwise identical anyway; with it on, pinning float32 would be the thing that breaks parity. Added a comment on that line so the rationale is visible.

File: src/maxtext/m3/core/rope.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  # (batch, length) -> (batch, length, 1, half_dim), broadcasting over heads.
+  angles = positions[:, :, jnp.newaxis, jnp.newaxis].astype(jnp.float32) / timescale
+  sin = jnp.sin(angles).astype(x.dtype)
+  cos = jnp.cos(angles).astype(x.dtype)
+
+  # Rotate each (x1[i], x2[i]) pair by its angle. Equivalent to the legacy
+  # `x * cos + rotate_half(x) * sin`, without materializing the doubled
+  # sin/cos tables.
+  x1, x2 = jnp.split(x, 2, axis=-1)
+  return jnp.concatenate([x1 * cos - x2 * sin, x2 * cos + x1 * sin], axis=-1)
```
[the remark this answers is not in the document]
Comment (bvandermoon) [reply src/maxtext/m3/core/rope.py]: Updated

**Review by khatwanimohit — APPROVED** [review]

**Review by hengtaoguo — APPROVED** [review]
